etna/records/models.py

```diff
@@ -1,78 +1,261 @@
+from .converters import IAIDConverter
+
+logger = logging.getLogger(__name__)
 
 
-@dataclass
 class Record(DataLayerMixin, APIModel):
-    """Non-creatable page used to render record data in templates.
-
-    This stub page allows us to use common templates to render external record
-    data as though the data was fetched from the CMS.
-
-    see: views.record_detail_view
-    """
-
-    iaid: str
-    title: str
-    reference_number: str
-
-    legal_status: str = ""
-    created_by: str = ""
-    description: str = ""
-    origination_date: str = ""
-    closure_status: str = ""
-    availability_delivery_condition: str = ""
-    arrangement: str = ""
-    held_by: str = ""
-    is_digitised: bool = False
-    parent: dict = field(default_factory=dict)
-    hierarchy: dict = field(default_factory=dict)
-    media_reference_id: str = ""
-    availability_delivery_surrogates: dict = field(default_factory=dict)
-    topics: dict = field(default_factory=dict)
-    next_record: dict = field(default_factory=dict)
-    previous_record: dict = field(default_factory=dict)
-    related_records: dict = field(default_factory=dict)
-    related_articles: dict = field(default_factory=dict)
-    catalogue_source: str = ""
-    repo_summary_title: str = ""
-    repo_archon_value: str = ""
-    level_code: str = ""
-    level: str = ""
-    template_reference_number: str = ""
-    template_summary_title: str = ""
-    hierarchy_level3_reference_number: str = ""
-    hierarchy_level3_summary_title: str = ""
-    related_materials: list = field(default_factory=list)
-
-    _debug_kong_result: dict = field(default_factory=dict)
+    """A 'lazy' data-interaction layer for record data retrieved from the Kong API"""
+
+    def __init__(self, raw_data: Dict[str, Any]):
+        """
+        This method recieves the raw JSON data dict recieved from
+        Kong and makes it available to the instance as `self._raw`.
+        """
+        self.score = raw_data.get("_score")
+        self._raw = raw_data.get("_source") or raw_data
+
+    @classmethod
+    def from_api_response(cls, response: dict) -> Record:
+        return cls(response)
 
     def __str__(self):
         return f"{self.title} ({self.iaid})"
 
-    def get(self, key: str, default: Optional[Any] = None):
-        return getattr(self, key, default)
+    def get(self, key: str, default: Optional[Any] = "", fail_loudly=False):
+        if "." in key:
+            return extract(self._raw, key, default, fail_loudly)
+        try:
+            return self._raw[key]
+        except KeyError as e:
+            if fail_loudly:
+                raise VariableDoesNotExist(str(e))
+            return default
 
-    @classmethod
-    def from_api_response(cls, response: dict) -> Record:
-        return cls(**transform_record_result(response))
+    @cached_property
+    def template(self) -> Dict[str, Any]:
+        return self.get("@template.details", self.get("@template.results", {}))
```

Comment:
1. Because, as you know, there are [two possible templates ](https://github.com/nationalarchives/ds-wagtail/blob/develop/etna/ciim/client.py#L51-L58) we can request. The `Record` class prefers the `details` template, because it's more complete. But, if that isn't available, the `results` template still usually contains the important stuff, hence why we fall back to that

Reply:
2. I'm not sure what we gain from knowing about the `template` param that was used. Either the template data is present or it isn't. Because this is a cached property, the work to figure out the return value is only done once... the first time it is requested. On every subsequent request to `self.template`, the cached return value is used.

Reply:
Its pointless keeping something like `@template.results` when 
1. Not sure what its used for,
2. `@template.results` is  also available in the response as the input param `template=results` is not used in the code.
3. Gives wrong impression that the results response is available.

Reply:
@TNA-Allan But it's not pointless... we know that it can be used and can be present in the representations. The ultimate fallback is `{}` (a blank dict), so it's okay if it isn't there.... but we can certainly use it if it is.

Reply:
> The ultimate fallback is {} (a blank dict)
The pointlessness is not about blank dict, also it was not in question.

Reply:
> but we can certainly use it if it is.
Surely, as long we use it when its needed, not otherwise.

Reply:
I'm honestly lost here @TNA-Allan. Can you articulate what you want to change and why?

Reply:
> I'm honestly lost here @TNA-Allan. Can you articulate what you want to change and why?
@ababic  The change is to remove @template.results and several reasons are mentioned in the my point above. Happy do discuss further if you're still lost.

Reply:
@TNA-Allan But I explained in my first reply about this:
> There [two possible templates](https://github.com/nationalarchives/ds-wagtail/blob/develop/etna/ciim/client.py#L51-L58) we can request from Kong

While this is true of the API, wouldn't you find it odd if you requested the `results` template, but the Record instance couldn't access that data, even if the API provided it?

Reply:
> wouldn't you find it odd if you requested the results template

See below if this helps why we do not want it-

> Not sure what its used for,
@template.results is available in the response as the input param template=results but this param is not used in the code.
Gives wrong impression that the results response is available.

Reply:
We could definitely use it whenever it's param is used in the code.

Reply:
Try to detach this code in your mind from how it is currently being used currently. You should see this as a generic interface for interacting with record data. While it's true that it makes the values easier to access in views, that is not the limit of it's usefulness... maybe you'll need to write a management command or write a new view soon that does use `template=results` - You shouldn't have to worry about whether that data will be respected or not at the time... you would just expect it to work.

Reply:
With respect, only catering for 'right now' is the very reason changes like this are needed. Sometimes you've got to pull your mind away from the immediate requirements and develop interfaces that make sense in their own right. When you allow two interfaces to bend too much to the whims of the other, you quickly end up in a mess.

Reply:
I differ on this at this time, but going with your reasoning.